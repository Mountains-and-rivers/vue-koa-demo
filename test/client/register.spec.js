import Vue from 'vue'
import VueRouter from 'vue-router'
import ElementUI from 'element-ui'
import { mount } from '@vue/test-utils'
import Register from '@/components/Register'

Vue.use(ElementUI)
Vue.use(VueRouter)
const router = new VueRouter()

jest.mock('axios', () => ({
  post: jest.fn()
    // for test4
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: false
      }
    }))
    // for test5
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
}))

describe('Register.vue', () => {
  let wrapper
  beforeEach(() => {
    jest.useFakeTimers()
    wrapper = mount(Register, { sync: false, router })
  })

  test('confirmRegister method should be called when clicking register button', async () => {
    const stub = jest.fn()
    wrapper.setMethods({ confirmRegister: stub })
    await wrapper.vm.$nextTick()
    wrapper.find('.btns .el-button:nth-child(2)').trigger('click')
    expect(stub).toBeCalledTimes(1)
  })

  test('register failed when all fields are empty', async () => {
    const result = await wrapper.vm.confirmRegister('form')
    expect(result).toBe(false)
  })

  test('register failed when password is not same as repeated password', async () => {
    wrapper.setData({
      form: {
        name: 'breezymelon',
        pwd: '111',
        repwd: '1111'
      }
    })
    const result = await wrapper.vm.confirmRegister('form')
    expect(result).toBe(false)
  })

  test('register failed when username has already exist', async () => {
    wrapper.setData({
      form: {
        name: 'breezymelon',
        pwd: '111',
        repwd: '111'
      }
    })
    const result = await wrapper.vm.confirmRegister('form')
    expect(result).toBe(false)
  })

  test('register success when all fields are correct', async () => {
    wrapper.setData({
      form: {
        name: 'melon',
        pwd: '111',
        repwd: '111'
      }
    })
    const result = await wrapper.vm.confirmRegister('form')
    jest.runAllTimers()
    expect(result).toBe(true)
  })

  test('register.vue snapshot test', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
