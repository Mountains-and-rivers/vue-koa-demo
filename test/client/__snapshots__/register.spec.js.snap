// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register.vue register.vue snapshot test 1`] = `
<div
  class="register"
>
  <form
    class="el-form"
  >
    <div
      class="el-form-item el-form-item--feedback is-required"
    >
      <label
        class="el-form-item__label"
        for="name"
        style="width: 80px;"
      >
        用户名
      </label>
      <div
        class="el-form-item__content"
        style="margin-left: 80px;"
      >
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            type="text"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="el-form-item el-form-item--feedback is-required"
    >
      <label
        class="el-form-item__label"
        for="pwd"
        style="width: 80px;"
      >
        密 码
      </label>
      <div
        class="el-form-item__content"
        style="margin-left: 80px;"
      >
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            type="password"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="el-form-item el-form-item--feedback is-required"
    >
      <label
        class="el-form-item__label"
        for="repwd"
        style="width: 80px;"
      >
        确认密码
      </label>
      <div
        class="el-form-item__content"
        style="margin-left: 80px;"
      >
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            type="password"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="btns"
    >
      <button
        class="el-button el-button--default"
        type="button"
      >
        <!---->
        <!---->
        <span>
          取消注册
        </span>
      </button>
       
      <button
        class="el-button el-button--primary"
        type="button"
      >
        <!---->
        <!---->
        <span>
          确认注册
        </span>
      </button>
    </div>
  </form>
</div>
`;
